File: backend/api/nodes.py
"""API endpoints for research node CRUD operations."""

from fastapi import APIRouter, HTTPException
from neomodel import DoesNotExist  # type: ignore[attr-defined]

from backend.schemas.nodes import NodeCreate, NodeResponse, NodeUpdate
from backend.services import node_service

router = APIRouter(prefix="/nodes", tags=["nodes"])


@router.post("", response_model=NodeResponse, status_code=201)
def create_node(data: NodeCreate) -> NodeResponse:
    return node_service.create_node(data)


@router.get("", response_model=list[NodeResponse])
def list_nodes(
    node_type: str | None = None,
    needs_review: bool | None = None,
    search_query: str | None = None,
    limit: int = 100,
    offset: int = 0,
) -> list[NodeResponse]:
    return node_service.list_nodes(
        node_type=node_type,
        needs_review=needs_review,
        search_query=search_query,
        limit=limit,
        offset=offset,
    )


@router.get("/{uid}", response_model=NodeResponse)
def get_node(uid: str) -> NodeResponse:
    try:
        return node_service.get_node(uid)
    except DoesNotExist:
        raise HTTPException(status_code=404, detail=f"Node '{uid}' not found") from None


@router.patch("/{uid}", response_model=NodeResponse)
def update_node(uid: str, data: NodeUpdate) -> NodeResponse:
    try:
        return node_service.update_node(uid, data)
    except DoesNotExist:
        raise HTTPException(status_code=404, detail=f"Node '{uid}' not found") from None


@router.delete("/{uid}", status_code=204)
def delete_node(uid: str) -> None:
    try:
        node_service.delete_node(uid)
    except DoesNotExist:
        raise HTTPException(status_code=404, detail=f"Node '{uid}' not found") from None

File: frontend/src/api/nodes.ts
import { api } from './client.ts';
import type { ResearchNodeData, NodeCreateData, NodeUpdateData } from '../types/node.ts';

export async function fetchNodes(params?: Record<string, string | boolean | number>): Promise<ResearchNodeData[]> {
  const { data } = await api.get('/nodes', { params });
  return data;
}

export async function createNode(payload: NodeCreateData): Promise<ResearchNodeData> {
  const { data } = await api.post('/nodes', payload);
  return data;
}

export async function updateNode(uid: string, payload: NodeUpdateData): Promise<ResearchNodeData> {
  const { data } = await api.patch(`/nodes/${uid}`, payload);
  return data;
}

export async function deleteNode(uid: string): Promise<void> {
  await api.delete(`/nodes/${uid}`);
}
